clues_test.go

```diff
@@ -683,3 +698,64 @@ func TestAddComment_trace(t *testing.T) {
 
 	commentMatches(t, expected, stack)
 }
+
+func TestAddWitness(t *testing.T) {
+	ctx := context.Background()
+	ctx = clues.Add(ctx, "one", 1)
+
+	mapEquals(t, ctx, msa{
+		"one": 1,
+	}, true)
+
+	ctxWithWit := clues.AddWitness(ctx, "wit")
+	clues.Relay(ctx, "wit", "zero", 0)
+	clues.Relay(ctxWithWit, "wit", "two", 2)
+
+	mapEquals(t, ctx, msa{
+		"one": 1,
+	}, true)
+
+	mapEquals(t, ctxWithWit, msa{
+		"one": 1,
+		"witnessed": map[string]map[string]any{
+			"wit": {
+				"two": 2,
+			},
+		},
+	}, true)
+
+	ctxWithTim := clues.AddWitness(ctxWithWit, "tim")
+	clues.Relay(ctxWithTim, "tim", "three", 3)
+
+	mapEquals(t, ctx, msa{
+		"one": 1,
+	}, true)
+
+	mapEquals(t, ctxWithTim, msa{
+		"one": 1,
+		"witnessed": map[string]map[string]any{
+			"wit": {
+				"two": 2,
+			},
+			"tim": {
+				"three": 3,
+			},
+		},
+	}, true)
+
+	ctxWithBob := clues.AddWitness(ctx, "bob")
+	clues.Relay(ctxWithBob, "bob", "four", 4)
+
+	mapEquals(t, ctx, msa{
+		"one": 1,
+	}, true)
+
+	mapEquals(t, ctxWithBob, msa{
```

Comment:
nit: could you please also make sure that the first two contexts with witnesses aren't changed?

Reply:
sure!